### PR TITLE
Fix issue 1043: Obfuscate AI data on request redirect

### PR DESCRIPTION
**Pull request by cristinamanum**
The AI data can contain user names. The user names are in general on the request url field but in request redirect case it can be found on the name and operation_Name fields as well.

With this change:
1. The redirect requests data will be obfuscated 
A name like: GET /profiles/cristinamanu will be changed in GET /profiles/ObfuscatedUserName 
An url like: http://localhost/profiles/cristinamanu will be changed in http://localhost/profiles/ObfuscatedUserName
A operation_Name like: GET /profiles/cristinamanu will be changed in GET /profiles/ObfuscatedUserName 
2. The not redirected requests will be obfuscated and kept in sync with the above structure.  A snapshot of the obfuscated values:
image

### CODE REVIEW

File: src/NuGetGallery/Telemetry/Obfuscator.cs [hunk 1]
```diff
@@ -25,9 +51,55 @@ internal static class Obfuscator
...
+            string obfuscatedValue = Regex.Replace(value,
+                                             ObfuscatedTemplates[obfuscatedTemplateKey].ObfuscateTemplate,
+                                             ObfuscatedTemplates[obfuscatedTemplateKey].ObfuscateValue);
+            return obfuscatedValue;
+        }
+
+        internal static bool NeedsObfuscation(string valueToBeVerified, out string obfuscatedTemplateKey)
+        {
+            obfuscatedTemplateKey = string.Empty;
+            var match = ObfuscatedTemplates.Where(template => Regex.IsMatch(valueToBeVerified.ToLower(), template.Key)).FirstOrDefault();
```
Comment (joelverhagen) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: Please add an `OrderBy` to the static so that order is defined. Dictionaries don't have a defined order AFAIK

File: src/NuGetGallery/Telemetry/Obfuscator.cs [hunk 2]
```diff
@@ -25,9 +51,55 @@ internal static class Obfuscator
             "Users/Profiles",
             "Users/ResetPassword"};
 
+        internal static readonly Dictionary<string, ObfuscateMetadata> ObfuscatedTemplates = new Dictionary<string, ObfuscateMetadata>
+        {
+            {@"/packages/(.+)/owners/(.+)/confirm/(.+)", new ObfuscateMetadata(@"/owners/(.+)/confirm", $"/owners/{DefaultTelemetryUserName}/confirm")},
```
Comment (joelverhagen) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: We should probably compile these regular expressions.

File: src/NuGetGallery/Telemetry/Obfuscator.cs [hunk 3]
```diff
@@ -25,9 +51,55 @@ internal static class Obfuscator
...
+            {@"/packages/(.+)/owners/(.+)/reject/(.+)", new ObfuscateMetadata(@"/owners/(.+)/reject", $"/owners/{DefaultTelemetryUserName}/reject") },
+            {@"/packages/(.+)/owners/(.+)/cancel/(.+)", new ObfuscateMetadata(@"/owners/(.+)/cancel", $"/owners/{DefaultTelemetryUserName}/cancel") },
+            {@"/account/confirm/(.+)/(.+)", new ObfuscateMetadata(@"/account/confirm/(.+)/", $"/account/confirm/{DefaultTelemetryUserName}/") },
+            {@"/account/delete/(.+)", new ObfuscateMetadata(@"/account/delete/(.+)", $"/account/delete/{DefaultTelemetryUserName}") },
+            {@"/profiles/(.+)", new ObfuscateMetadata(@"/profiles/(.+)", $"/profiles/{DefaultTelemetryUserName}") },
+            {@"/account/setpassword/(.+)/(.+)", new ObfuscateMetadata(@"/setpassword/(.+)/", $"/setpassword/{DefaultTelemetryUserName}/") },
+            {@"/account/forgotpassword/(.+)/(.+)", new ObfuscateMetadata(@"/forgotpassword/(.+)/", $"/forgotpassword/{DefaultTelemetryUserName}/") }
```
Comment (joelverhagen) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: Why `@`?

File: src/NuGetGallery/Telemetry/Obfuscator.cs [hunk 4]
```diff
@@ -25,9 +51,55 @@ internal static class Obfuscator
...
+        }
+
+        internal static string Obfuscate(string value)
+        {
+            string obfuscatedTemplateKey = null;
+            if (value == null || !NeedsObfuscation(value, out obfuscatedTemplateKey))
+            {
+                return value;
+            }
+            string obfuscatedValue = Regex.Replace(value,
```
Comment (joelverhagen) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: Is this case sensitive? I think our URLs are:
https://www.nuget.org/PROFILES/jamesnk

File: src/NuGetGallery/Telemetry/Obfuscator.cs [hunk 5]
```diff
@@ -25,9 +51,55 @@ internal static class Obfuscator
...
             return url == null ? string.Empty : $"{url.Scheme}://{url.Host}/{DefaultTelemetryUserName}";
         }
+
+        internal static Uri ObfuscateUrl(Uri url)
+        {
+            if (url == null)
+            {
+                return url;
+            }
+            return new Uri($"{url.Scheme}://{url.Host}{Obfuscate(url.AbsolutePath)}");
```
Comment (joelverhagen) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: Do we lose port?

Comment (joelverhagen) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: (e.g. if a customer is running gallery on a non-standard port)

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: Use case insensitive regex instead of `ToLower()`

**Comment by chenriksson** [comment]
@cristinamanum string regex for obfuscating urls seems fragile and error-prone. Would it be possible instead to just use the existing request's RouteData? Maybe we could even inject logic when mapping routes to specify what should be obfuscated?

Very rough idea of what I'm thinking...
```
class Routes {
    routes.MapRoute(
        name: RouteName.Profile,
        url: "profiles/{username}",
        defaults: new { controller = "Users", action = "Profiles" },
        obfuscations: new { username = 'ObfuscatedUsername' }
}

static class NuGetRouteExtensions {

    public static Dictionary<string, Dictionary<string, string>> UrlObfuscations;

    // using Dictionary, or could use object ViewBag to be consistent with MVC routing APIs
    public static Route MapRoute(this RouteCollection routes, string name, string url, object defaults, Dictionary<string, string> obfuscations) {
        if (obfuscations != null) {
            Obfuscator.UrlObfuscations[url] = obfuscations;
        }
    }

    public static string ObfuscateUri(this Route route, string actualUri) {
        if (UrlObfuscations.ContainsKey(route.Uri) {
            var obfuscations = UrlObfuscations[route.Uri];
            var parts = routeUri.Split('/');
            var actualParts = actualUri.Split('/');
            for (int i = 0; i < parts.length; i++) {
                if (obfuscations.ContainsKey(parts[i])) {
                    actualParts[i] = obfuscations[parts[i]];
                }
            }
            return actualParts.Join('/');
        }
    }
}
```

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 2 above)
Comment (loic-sharma) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: Do we already use compiled regex expressions in the gallery? .NET only caches up to 15 compiled regex expressions by default (source [here](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.cachesize?view=netcore-2.0#System_Text_RegularExpressions_Regex_CacheSize)), and compiling a regex is rather expensive.

TLDR: Compiling regex may result in worse performance, we should due our due diligence before turning it on.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 4 above)
Comment (skofman1) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: Add timeouts to all regex matches: https://github.com/NuGet/NuGetGallery/issues/4038

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: Nice read, good point.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 1 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: Updated to use case insensitive. 
Why to orderby? The return is only one element.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 2 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: I did simple perf tests with: static regex and compiled instance regex . The results show similar perf characteristics. 
For invocation of 1000 https://localhost/profiles/cristinamanu
        Static - 525.6479932 seconds
	Instance Compiled - 525.0719887 seconds

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 3 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: I tend to use verbatim string when working with regex expressions. For these regex there is no difference.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 4 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: I added case insensitive option.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 5 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: Added the port.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 4 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: Added the timeouts.

File: src/NuGetGallery/Extensions/RouteExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        public static void MapRoute(this RouteCollection routes, string name, string url, object defaults, ObfuscatedMetadata obfuscationMetadata)
+        {
+            routes.MapRoute(name, url, defaults);
+            if (!ObfuscatedRouteMap.ContainsKey(url)) { ObfuscatedRouteMap.Add(url, obfuscationMetadata); }
+        }
+
+        public static string ObfuscateUrlPath(this Route route, string urlPath)
+        {
+            var path = route.Url;
```
Comment (chenriksson) [inline src/NuGetGallery/Extensions/RouteExtensions.cs]: nit: extra single whitespace?

File: src/NuGetGallery/Extensions/RouteExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Web.Mvc;
+using System.Web.Routing;
+
+namespace NuGetGallery
+{
+    public static class RouteExtensions
```
Comment (chenriksson) [inline src/NuGetGallery/Extensions/RouteExtensions.cs]: nit: class could be internal (tests already have internalsVisibleTo)

File: src/NuGetGallery/Extensions/RouteExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            { get; }
+
+            public ObfuscatedMetadata(int obfuscatedSegment, string obfuscateValue)
+            {
+                ObfuscatedSegment = obfuscatedSegment;
+                ObfuscateValue = obfuscateValue;
+            }
+        }
+
+        internal static Dictionary<string, ObfuscatedMetadata> ObfuscatedRouteMap = new Dictionary<string, ObfuscatedMetadata>();
```
Comment (chenriksson) [inline src/NuGetGallery/Extensions/RouteExtensions.cs]: nit: maybe name `RoutePartsToObfuscateByUrl` is clearer?

File: src/NuGetGallery/Extensions/RouteExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var path = route.Url;
+           if (!ObfuscatedRouteMap.ContainsKey(path))
+            {
+                return null;
+            }
+            var metadata = ObfuscatedRouteMap[path];
+            string[] segments = urlPath.Split('/');
+            segments[metadata.ObfuscatedSegment] = metadata.ObfuscateValue;
+            return segments.Aggregate((x, y) => { return x + "/" + y; });
```
Comment (chenriksson) [inline src/NuGetGallery/Extensions/RouteExtensions.cs]: Is this equivalent to `String.Join('/', segments)`? Is Join simpler?

File: src/NuGetGallery/Helpers/RegexEx.cs [hunk 10]
```diff
@@ -8,7 +8,7 @@ namespace NuGetGallery.Helpers
 {
     public static class RegexEx
     {
-        private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(2);
+        internal static readonly TimeSpan Timeout = TimeSpan.FromSeconds(2);
```
Comment (chenriksson) [inline src/NuGetGallery/Helpers/RegexEx.cs]: still necessary?

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs [hunk 11]
```diff
@@ -30,29 +30,32 @@ private void ModifyItem(ITelemetry item)
...
 
-        private Uri ObfuscateUri(RequestTelemetry telemetryItem)
+        public virtual Route Route
         {
-            if(IsPIIOperation(telemetryItem.Context.Operation.Name))
+            get
             {
-                // The new url form will be: https://nuget.org/ObfuscatedUserName
-                return new Uri(Obfuscator.DefaultObfuscatedUrl(telemetryItem.Url));
+                return RouteTable.Routes.GetRouteData(new System.Web.HttpContextWrapper(HttpContext.Current)).Route as Route;
```
Comment (chenriksson) [inline src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: nit: namespace unnecessary since 'System.Web' is imported?
nit: maybe this is complex enough to be method like `GetCurrentRouteData`?

File: src/NuGetGallery/Telemetry/Obfuscator.cs [hunk 12]
```diff
@@ -2,8 +2,10 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections;
 using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using NuGetGallery.Helpers;
```
Comment (chenriksson) [inline src/NuGetGallery/Telemetry/Obfuscator.cs]: nit: can revert?

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs [hunk 13]
```diff
@@ -30,29 +30,32 @@ private void ModifyItem(ITelemetry item)
...
         }
 
-        private Uri ObfuscateUri(RequestTelemetry telemetryItem)
+        public virtual Route Route
         {
-            if(IsPIIOperation(telemetryItem.Context.Operation.Name))
+            get
             {
-                // The new url form will be: https://nuget.org/ObfuscatedUserName
-                return new Uri(Obfuscator.DefaultObfuscatedUrl(telemetryItem.Url));
```
Comment (chenriksson) [inline src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: DefaultObfuscatedUrl was also called by QuietLog.GetObfuscatedServerVariables. Can this use the same technique?

**Comment by chenriksson** [comment]
@cristinamanum Looks really nice!

File: src/NuGetGallery/Extensions/RouteExtensions.cs (hunk 8 above)
Comment (cristinamanum) [reply src/NuGetGallery/Extensions/RouteExtensions.cs]: I think I like more the current name.

File: src/NuGetGallery/Extensions/RouteExtensions.cs (hunk 9 above)
Comment (cristinamanum) [reply src/NuGetGallery/Extensions/RouteExtensions.cs]: Yep - join has better perf than aggregate - updated.

File: src/NuGetGallery/Extensions/RouteExtensions.cs (hunk 6 above)
Comment (cristinamanum) [reply src/NuGetGallery/Extensions/RouteExtensions.cs]: updated

File: src/NuGetGallery/Helpers/RegexEx.cs (hunk 10 above)
Comment (cristinamanum) [reply src/NuGetGallery/Helpers/RegexEx.cs]: no - updated

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs (hunk 11 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: Comment 1 : it was not needed - updated 
Comment 2 : I agree method is better suited - updated

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs (hunk 13 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: It will be updated if needed in a different change.

File: src/NuGetGallery/Telemetry/Obfuscator.cs (hunk 12 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/Obfuscator.cs]: yes - updated

File: src/NuGetGallery/Extensions/RouteExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+                ObfuscateValue = obfuscateValue;
+            }
+        }
+
+        internal static Dictionary<string, ObfuscatedMetadata> ObfuscatedRouteMap = new Dictionary<string, ObfuscatedMetadata>();
+
+        public static void MapRoute(this RouteCollection routes, string name, string url, object defaults, ObfuscatedMetadata obfuscationMetadata)
+        {
+            routes.MapRoute(name, url, defaults);
+            if (!ObfuscatedRouteMap.ContainsKey(url)) { ObfuscatedRouteMap.Add(url, obfuscationMetadata); }
```
Comment (skofman1) [inline src/NuGetGallery/Extensions/RouteExtensions.cs]: nit: { should be in a new line

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs [hunk 15]
```diff
@@ -30,29 +30,29 @@ private void ModifyItem(ITelemetry item)
...
             if(requestTelemetryItem != null)
             {
-                requestTelemetryItem.Url = ObfuscateUri(requestTelemetryItem);
+                var route = GetCurrentRoute();
+                if(route == null)
+                {
+                    return;
+                }
+                // Removes the first /
+                var requestPath = requestTelemetryItem.Url.AbsolutePath.Remove(0,1);
```
Comment (shishirx34) [inline src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: probably a good idea to test if the first char is '/' before removal?

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs [hunk 16]
```diff
@@ -30,29 +30,29 @@ private void ModifyItem(ITelemetry item)
...
+                var obfuscatedPath = route.ObfuscateUrlPath(requestPath);
+                if(obfuscatedPath != null)
+                {
+                    requestTelemetryItem.Url = new Uri(requestTelemetryItem.Url.ToString().Replace(requestPath, obfuscatedPath));
+                    requestTelemetryItem.Name = requestTelemetryItem.Name.Replace(requestPath, obfuscatedPath);
+                    if(requestTelemetryItem.Context.Operation != null && requestTelemetryItem.Context.Operation.Name != null)
```
Comment (shishirx34) [inline src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: you can simplify this: `if (requestTelemetryItem.Context.Operation?.Name != null)`

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs (hunk 15 above)
Comment (chenriksson) [reply src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: Maybe use `requestTelemetryItem.Url.AbsolutePath.TrimStart('/')`?

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs (hunk 16 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: nicer - updated

File: src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs (hunk 15 above)
Comment (cristinamanum) [reply src/NuGetGallery/Telemetry/ClientTelemetryPIIProcessor.cs]: I tested
updated  with TrimStart
